`ActiveStorage` immediate variants

tomrossi7 opened the pull request:
### Motivation / Background

ActiveStorage currently allows variants to be created on demand (by default) or in the background (using preprocessed). This can be problematic for applications that have a lot of traffic. Immediately after an attachment is created, thousands of requests could be created for a single variant. Each controller receiving the request as well as a background job could all be racing to create the variant.

In order to give more control variant creation, we are proposing an immediate option to allow variants to be created simultaneously with the attachment.

This Pull Request has been created because products like ours need more control over variant creation.

### Detail

This Pull Request changes introduces the `immediate` option and follows the existing patterns around `preprocessed`.

```
  has_one_attached :avatar_with_immediate do |attachable|
    attachable.variant :thumb, resize_to_limit: [4, 4], immediate: true
  end
```

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

p8 commented:
Thanks for the PR @tomrossi7 !
This seems very similar to preprocessed variants:
https://github.com/rails/rails/pull/47473

p8 commented:
Maybe we should have non boolean values for preprocessed?
`preprocessed: :later`
`preprocessed: :immediate`
Or maybe :immediate should be the default?

zzak commented:
What happens if you have both `preprocessed` and `immediate` options?

I think @p8's suggestion is reasonable, but I imagine this adds a bit more complexity than we're willing to take on, maybe it's better to have an explicit option like this so the developer can clearly tell the difference.

tomrossi7 commented:
> What happens if you have both `preprocessed` and `immediate` options? 
It will create the variants immediately and then the background job will see that they've already been created. It won't error out, but it isn't efficient.

> I think @p8's suggestion is reasonable, but I imagine this adds a bit more complexity than we're willing to take on, maybe it's better to have an explicit option like this so the developer can clearly tell the difference.
I would prefer it to be explicit as well, but currently, the `preprocessed` option [accepts a bolean, a symbol, or a proc](https://github.com/rails/rails/blob/f9644ca3f76cc92ec26b3f86ea9a73f7d2980efa/activestorage/app/models/active_storage/named_variant.rb#L22). We couldn't change that without it being a breaking change.

MatheusRich reviewed (commented):
Do we need to update the guides?

tomrossi7 commented:
> Do we need to update the guides?

I'm happy to update the guides. I just wanted to make sure we are good with the changes.

tomrossi7 commented:
Is there anything else I can do here to help this PR along?

There are sections of the original code that could be refactored, but I didn't want to introduce too much change at once.

HamptonMakes commented:
I was also caught off guard by this change and was expecting the behavior @tomrossi7 is describing!

Would very much like this feature.

ConfusedVorlon commented:
> > What happens if you have both `preprocessed` and `immediate` options?
> 
> It will create the variants immediately and then the background job will see that they've already been created. It won't error out, but it isn't efficient.
> 
> > I think @p8's suggestion is reasonable, but I imagine this adds a bit more complexity than we're willing to take on, maybe it's better to have an explicit option like this so the developer can clearly tell the difference.
> 
> I would prefer it to be explicit as well, but currently, the `preprocessed` option [accepts a bolean, a symbol, or a proc](https://github.com/rails/rails/blob/f9644ca3f76cc92ec26b3f86ea9a73f7d2980efa/activestorage/app/models/active_storage/named_variant.rb#L22). We couldn't change that without it being a breaking change.

is there an argument to create a new option that contains both behaviours - so that ultimately `preprocessed:` can be deprecated?

so, perhaps 
```
creation: :immediate
creation: :background
creation: :on_demand #default
```

preprocessed would initially just set `creation: :background`, and some point in the future, it could be deprecated.

tenderlove commented:
This feature makes sense to me, but I'd like to take a little more time to read the option names. I like @ConfusedVorlon's suggestion (but as I said, I need to reread)

searls commented:
👍 to @ConfusedVorlon's proposal. A single, more clearly-named option that covers each of the modes. As someone who's been using AS in anger for six months, the default behavior is surprising and `preprocessed` is indeed overselling what happens. In fact, after thinking for a moment, the entire reason I switched all my variants to `preprocessed` was to avoid this exact risk, but it would seem that `on_demand` is the better option.

As for the feature itself, it does have me thinking about extremely large files. If someone direct-uploads a 20GB file to AWS, then the puma worker process has to download the whole thing and pipe it to ffmpeg to generate a preview image, right? If so, then it'd only take a few immediate/on-demand variants to be run at once to consume all available processes for many users.

tenderlove commented:
@searls in worried about the DoS aspect as well. I have to imagine this mode would be for CRM type apps where the user is somewhat trusted. Or the proxy prevents uploads that big (or something). Either way it seems like a footgun that should be documented (if it isn't already)

tomrossi7 commented:
Right now the `preprocessed` option is doing double duty. The option is specifying whether or not it should be processed in the background as well as optionally the custom transformation method name or proc to execute. We could break this into two separate options the `creation` option to specify when it should be created and optionally `custom_transformation` for a method name or proc to be executed.


```
  has_one_attached :avatar_with_immediate do |attachable|
    attachable.variant :thumb, resize_to_limit: [4, 4], creation: :immediate
  end

  has_one_attached :avatar_with_immediate do |attachable|
    attachable.variant :thumb, creation: :immediate, custom_transformation: :my_custom_method
  end

  has_one_attached :avatar_with_immediate do |attachable|
    attachable.variant :thumb, creation: :immediate, custom_transformation: -> {...}
  end

```

tenderlove commented:
> Right now the preprocessed option is doing double duty. The option is specifying whether or not it should be processed in the background as well as optionally the custom transformation method name or proc to execute.

Specifying "how" and "when" separately makes sense. We just have to ensure backwards compatibility 😁

tomrossi7 commented:
For backwards compatibility, I could put a deprecation warning and continue to respect the `preprocessed` option, but the use of the new options would override it. This with updated documentation may be a good solution?

tenderlove commented:
> For backwards compatibility, I could put a deprecation warning and continue to respect the `preprocessed` option, but the use of the new options would override it. This with updated documentation may be a good solution?

Yes, this sounds great!

tomrossi7 commented:
Oops! I misspoke. Currently the `preprocessed` option accepts a block or a method, but only to determine if the variant should be preprocessed:

```
has_one_attached :avatar_with_immediate do |attachable|
  attachable.variant :thumb, resize_to_limit: [4, 4], preprocessed: :should_preprocess?
end
```

If we wanted to continue this pattern of allowing a method to determine how the variant should be created, we would want to do something like this:

```
has_one_attached :avatar_with_immediate do |attachable|
  # when_to_create returns [:immediate, :delayed, or :dynamic]
  attachable.variant :thumb, resize_to_limit: [4, 4], creation: :when_to_create
end

has_one_attached :avatar_with_immediate do |attachable|
  # when_to_create returns [:immediate, :delayed, or :dynamic]
  attachable.variant :thumb, resize_to_limit: [4, 4], creation: -> {|object| object.when_to_create}
end
```
However, I don't love that this requires a method with special knowledge of approved return values. I would rather only support it explicitly for now.

I propose we deprecate the current `preprocessed` option and introduce a new `creation` option that explicitly accepts `immediate`, `delayed`, or `dynamic`:

```
has_one_attached :avatar_with_immediate do |attachable|
  # creation options are :immediate, :delayed, or :dynamic
  attachable.variant :thumb, resize_to_limit: [4, 4], creation: :immediate
end
```

tomrossi7 commented:
@tenderlove you and I spoke at Rails World about this PR. Do you like the API I've proposed above?

tenderlove commented:
> @tenderlove you and I spoke at Rails World about this PR. Do you like the API I've proposed above?

Yes, I think the API is great.  Just to check, would the `creation:` option have a default?  I think it should.

> However, I don't love that this requires a method or block with special knowledge of approved values. I would rather only support it explicitly for now.

I think this makes sense. I'm not sure when you'd need this to be calculated dynamically. I suppose we could do add support for a proc or something in the future if there's a use case.  I'm also generally in favor of fewer options.

tenderlove commented:
> Sure thing! I'll make it default to `:dynamic` since that is the current default behavior.

Great, thank you.  lmk when you're ready and I'll review again.  Thanks!

tomrossi7 commented on activestorage/app/jobs/active_storage/create_variants_job.rb:
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class ActiveStorage::CreateVariantsJob < ActiveStorage::BaseJob
```
The `CreateVariantsjob` is responsible for:
1. Determining when a preview is necessary for creating variants and generating it.
2. Either scheduling `TransformJob`s for each transformation or executing them immediately, depending on the generation strategy.

tomrossi7 commented on activestorage/app/jobs/active_storage/create_variants_job.rb:
```diff
@@ -0,0 +1,28 @@
+
+  private
+    def perform_method
+      (@generate == :immediate) ? :perform_now : :perform_later
```
The job uses the `@generate` to determine whether it needs to perform the `TransformJob`s now or later. I couldn't find a way for the job itself to know whether it is being run with a `perform_now` or a `perform_later`, so I passed it explicitly to the job.

tomrossi7 commented on activestorage/app/jobs/active_storage/transform_job.rb:
```diff
@@ -3,10 +3,10 @@
   queue_as { ActiveStorage.queues[:transform] }
 
-  discard_on ActiveRecord::RecordNotFound, ActiveStorage::UnrepresentableError
+  discard_on ActiveRecord::RecordNotFound
```
`TransformJob` is only called from one place and should not be called with an unrepresentable blob. Rather than discarding, I think it should raise an exception if someone were to break this.

tomrossi7 commented on activestorage/app/models/active_storage/named_variant.rb:
```diff
@@ -3,19 +3,26 @@
   end
+
+  private
+    def preprocessed?(record)
```
The preprocessed option continues to work and we could add a deprecation warning at any point.

tomrossi7 commented:
## Variant Generation Strategies

- **`:delayed`** - Variants are created in the background after attachment.
- **`:on_demand`** - Variants are created dynamically when requested.
- **`:immediate`** - Variants are created simultaneously with the attachment.

```ruby
# Delayed variants: created in the background after attachment
has_one_attached :avatar_with_delayed_variants do |attachable|
  attachable.variant :delayed_thumb, resize_to_limit: [2, 2], generation: :delayed
end

# On-demand variants: created dynamically when requested
has_one_attached :avatar_with_on_demand_variants do |attachable|
  attachable.variant :on_demand_thumb, resize_to_limit: [3, 3], generation: :on_demand
  attachable.variant :default_thumb, resize_to_limit: [4, 4]
end

# Immediate variants: created simultaneously with the attachment
has_one_attached :avatar_with_immediate_variants do |attachable|
  attachable.variant :immediate_thumb, resize_to_limit: [1, 1], generation: :immediate
end
```
@tenderlove let me know if I've gone too far!

This update introduces several changes designed to enhance the clarity, maintainability, and testability of the variant generation process in Active Storage.

A key part of this refactor is the introduction of the `CreateVariantsJob` class, which encapsulates the responsibility for creating variants either immediately or delayed. This job determines when a preview is necessary and handles its generation. It also either schedules `TransformJob`s for each transformation or executes them immediately, depending on the generation strategy. By moving this logic into a dedicated job, I was able to simplify the `Representable` module, which no longer contains code specific to variant creation. This change allows the `Representable` module to stay focused on its core purpose, making it more reusable and easier to maintain.

I also simplified the tests by concentrating on the `Attachment` class rather than `Attached::One` and `Attached::Many`, as the logic is the same for both.

Additionally, the `NamedVariant` class now determines the generation strategy for variants. This change encapsulates decision-making logic within the class that already has the most context about individual variants. This reduces some of the complexity required in the `Attachment` class to schedule `CreateVariantsJob` for each variant. 

Please tell me what you think and if you think I should go in a different direction!

tomrossi7 commented on activestorage/test/jobs/create_variants_job_test.rb:
```diff
@@ -0,0 +1,45 @@
+    ActiveStorage::CreateVariantsJob.perform_now blob, generate: :delayed, transformations: @transformations
+
+    @transformations.each do |transformation|
+      assert_not blob.variant(transformation).send(:processed?)
```
I didn't include in this PR, but I think we should consider moving `processed?` to a public method. Not only is it used in testing, but our app uses it as a conditional to determine whether to display a variant on highly requested pages.

searls replied:
It seems really weird there's no public way to ask if a variant is processed. I know in Build With Becky I wound up doing something like rescuing an unprocessed error instead

tomrossi7 replied:
Do you think the processed? Method should be public?

tomrossi7 commented:
@tenderlove can you give me some feedback this PR we've talked about. I'm happy to add documentation and possibly a deprecation warning, but want to make sure you like the overall direction.

rafaeldev commented:
@tomrossi7 does it impact the behavior about direct upload?

tomrossi7 commented:
> @tomrossi7 does it impact the behavior about direct upload?

Nope! This is behavior that kicks in after the direct upload has completed and the attachment is persisted. Let me know if you have any other questions!

Cosmo commented:
Sorry for bikeshedding, but I've been following this discussion and wanted to throw a naming alternative into the ring that aligns more closely with existing Rails naming conventions -- specifically from Active Job (`perform_now` and `perform_later`).

```ruby
# Process variant in the background
has_one_attached :avatar do |attachable|
  attachable.variant :thumb, resize_to_limit: [100, 100], processing: :later
end

# Process variant only when requested
has_one_attached :avatar do |attachable|
  attachable.variant :thumb, resize_to_limit: [200, 200], processing: :on_demand
end

# Process variant immediately on attachment
has_one_attached :avatar do |attachable|
  attachable.variant :thumb, resize_to_limit: [300, 300], processing: :now
end
```

I know `creation:` with `:immediate`, `:delayed` and `:dynamic` is currently the favored direction and it's a solid and clear design.  
That said, I personally lean toward naming harmony across Rails components and the `processing:` approach might feel more intuitive for developers who are familiar with Active Job.

ConfusedVorlon commented:
> ```ruby
> # Process variant in the background
> has_one_attached :avatar do |attachable|
>   attachable.variant :thumb, resize_to_limit: [100, 100], processing: :later
> end
> 
> # Process variant only when requested
> has_one_attached :avatar do |attachable|
>   attachable.variant :thumb, resize_to_limit: [200, 200], processing: :on_demand
> end
> 
> # Process variant immediately on attachment
> has_one_attached :avatar do |attachable|
>   attachable.variant :thumb, resize_to_limit: [300, 300], processing: :now
> end
> ```

This is nice. 

Still hoping this gets merged eventually :)

dhh commented:
This is great. I agree with Cosmo, though, that we can align the API closer with the existing vocabulary we have around jobs. A slight alteration on his proposal:

```ruby
# Process variant in the background
has_one_attached :avatar do |attachable|
  attachable.variant :thumb, resize_to_limit: [100, 100], process: :later
end

# Process variant only when requested
has_one_attached :avatar do |attachable|
  attachable.variant :thumb, resize_to_limit: [200, 200], process: :lazily
end

# Process variant immediately on attachment
has_one_attached :avatar do |attachable|
  attachable.variant :thumb, resize_to_limit: [300, 300], process: :immediately
end
```

The reason I like `immediately` over `now` is that we're this is setting up rules for invocation later, not calling a method that triggers in the moment, like `deliver_now` would. 

cc @jeremy

tomrossi7 committed:
Active storage immediate variants (#6)

* This commit introduces variant generation strategies for Active Storage,
providing more flexible control over when variants are created.

- **:lazily** - variants created dynamically when requested (default)
- **:later** - variants created in background after attachment
- **:immediately** - variants created simultaneously with attachment

```ruby
has_one_attached :avatar_with_lazy_variants do |attachable|
  attachable.variant :lazy_thumb, resize_to_limit: [3, 3], process: :lazy
  attachable.variant :default_thumb, resize_to_limit: [4, 4]
end

has_one_attached :avatar_with_later_variants do |attachable|
  attachable.variant :later_thumb, resize_to_limit: [2, 2], process: :later
end

has_one_attached :avatar_with_immediate_variants do |attachable|
  attachable.variant :immediate_thumb, resize_to_limit: [1, 1], process: :immediately
end
```

tomrossi7 commented:
Thanks for the feedback @dhh, I've gone ahead and updated the API to match the vocabulary proposed. If you and @jeremy approve, I can update the documentation as well. I also had a couple questions, I would love to get some feedback on:

1. Should we put a deprecation message on the `preprocessed` option? The option is still supported, but the new `process` option is more explicit and flexible.

2. Is it okay to make the `Variant#processed?` method public?

Here is the new API:

* This PR introduces variant generation strategies for Active Storage,
providing more flexible control over when variants are created.

- **:lazily** - variants created dynamically when requested (default)
- **:later** - variants created in background after attachment
- **:immediately** - variants created simultaneously with attachment

```ruby
has_one_attached :avatar_with_lazy_variants do |attachable|
  attachable.variant :lazy_thumb, resize_to_limit: [3, 3], process: :lazy
  attachable.variant :default_thumb, resize_to_limit: [4, 4]
end

has_one_attached :avatar_with_later_variants do |attachable|
  attachable.variant :later_thumb, resize_to_limit: [2, 2], process: :later
end

has_one_attached :avatar_with_immediate_variants do |attachable|
  attachable.variant :immediate_thumb, resize_to_limit: [1, 1], process: :immediately
end
```

tomrossi7 committed:
Merging with main

jeremy commented:
Yeah - this is great @tomrossi7. Giving it a spin locally.

jeremy committed:
Deprecate `preprocessed: true` in favor of `process: :later`

For removal in Rails 9.0.

Co-authored-by: Tom Rossi <tomrossi7@users.noreply.github.com>

jeremy committed:
Active Storage: make Variant#processed? and VariantWithRecord#processed? public

Apps can now check whether variants have been generated. Useful for
precise control of variant usage or generation.

Co-authored-by: Tom Rossi <tomrossi7@users.noreply.github.com>

jeremy committed:
Document immediate variants: guide and changelog

Co-authored-by: Tom Rossi <tomrossi7@users.noreply.github.com>

jeremy commented:
@tomrossi7 Validated locally and with Basecamp. Nice.

I like making `#processed?` public as part of this; would give it a separate changelog entry.

Made the last few changes but couldn't push to the branch. Feel free to use as a starting point: [active-storage.patch](https://github.com/user-attachments/files/23158563/active-storage.patch)

tomrossi7 committed:
Merge branch 'main' into active-storage-immediate-variants

p8 commented on activestorage/app/models/active_storage/attachment.rb:
```diff
@@ -131,20 +131,23 @@ def mirror_blob_later
-          named_variant.transformations
-        end
-      }
+    def create_variants
```
Maybe rename this to make clear variants might not get created immediately?
```suggestion
    def schedule_variant_transformations
```

tomrossi7 replied:
> Maybe rename this to make clear variants might not get created immediately?

Yeah, really it would be `create_or_schedule_create_variants`. My thinking was the method `create_variants` creates the variants based on their own creation strategy. 🤷‍♂️

tomrossi7 committed:
Making ActiveStorage::Preview#processed? a public method

Removing space

tomrossi7 committed:
Use consistent naming for transformations parameter (#7)

ActiveStorage uses the plural `transformations` for a single hash of transformation options (e.g., `{ resize_to_limit: [100, 100] }`).

To differentiate between a single transformation hash and an array of transformation hashes, this commit introduces `transformations_array` for arrays of transformations, while retaining transformations for individual hashes.

tomrossi7 committed:
Fixing changelog

tomrossi7 commented:
@jeremy let me know if you want me to make any additional changes.

tomrossi7 commented on activestorage/app/models/active_storage/attachment.rb:
```diff
@@ -131,20 +131,23 @@ def mirror_blob_later
+      named_variants.each do |_name, named_variant|
+        case named_variant.process(record)
+        when :immediately
+          immediate_transformations << named_variant.transformations
```
@p8 check out my updates when you get a chance!

jeremy reviewed (commented):
`transformations_array` feels clunky. It'd feel natural to accept either a single transformation hash or an array of them.

tomrossi7 commented on activestorage/app/jobs/active_storage/create_variants_job.rb:
```diff
@@ -6,14 +6,14 @@ class ActiveStorage::CreateVariantsJob < ActiveStorage::BaseJob
   retry_on ActiveStorage::IntegrityError, attempts: 10, wait: :polynomially_longer
 
-  def perform(blob, transformations:, process:)
+  def perform(blob, transformations_array:, process:)
```
> transformations_array feels clunky. It'd feel natural to accept either a single transformation hash or an array of them.

Maybe I can just use `transformations_array` with in the job to bury the clunkiness from everyone else? We could also call it something else `transformations_list`? I just think we need a name for the collection of hash transformations.

p8 replied:
Variants (the regular ones, not the named ones) are described as "the result of a set of transformations applied to the original."
https://github.com/rails/rails/blob/9f466dd9d4672d7dc6c49a7861d9e30eff69c163/activestorage/app/models/active_storage/variant.rb#L5
So maybe use `variants` instead of `transformations` or `transformations_array`?
It also lines up with the name of the job `CreateVariantsJob`.

tomrossi7 replied:
I definitely like it better than `transformations_array`!

tomrossi7 committed:
Rename transformations_array to variants

Aligns with `CreateVariantsJob` naming and `ActiveStorage` terminology
where variants represent the result of transformations applied to
the original blob.

Fixing create_preview_image

tomrossi7 commented:
Anything else I can do to help this along @jeremy @p8?
